_posts/2015-12-13-testing-multithreaded-code-in-java.md

```diff
@@ -0,0 +1,154 @@
+
+You can use a more deterministic approach using [Weaver](https://github.com/google/thread-weaver). To understand how it works let's illustrate it with an example. Let's say that we have an in-memory and not thread-safe store:
+
+```
+	private final Map<Level, Scores> scoresByLevel;
+```
+
+We have some service that access to a repo that wraps that store:
+
+```
+  1	Optional<Scores> scoresFromStore = scoreRepo.findBy(score.level());
+	2       if(scoresFromStore.isPresent()){
+	3          scoreRepo.update(score.level(), score);
+	4       }else{
+	5          scoreRepo.save(score.level(), new Scores().add(score));
+	6       }
+```
+
+That service is a singleton living in a server that spawns a thread per request, so we'd like to execute that piece atomically. We could use the stress test non-deterministic approach or we could use Weaver. If we think deeply in this problem, what we want to test is every combination of the following (as an example, Thread 1 executes line 1 in moment x and Thread 2 executes line 1 in moment x, would be -> T1/1 : T2/1)
+
+* T1/1 : T2/1
+* T1/1 : T2/2
+* T1/1 : T2/3
+* ....
+* T1/2 : T2/1
+* T1/2 : T2/2
+* T1/2 : T2/3
+* ....
+
+For instance, we'll have a problem if T1/5 and T2/2, as T1 didn't save yet, and T2 has already got an empty score from store. That means that T1 will save a score in a level and then T2 will do the same, breaking the logic. And that's what Weaver exactly does, it grabs a method and execute the previous combinations by two threads.
+
+If I get rid of preparation code (annotated with @ThreadedBefore), the test code will look like this:
+
+```
+    @ThreadedMain
+    public void mainThread() {
+        scoreService.save(LEVEL_ID, SCORE_VALUE, aUser);
+    }
+
+    @ThreadedSecondary
+    public void secondThread() {
+        scoreService.save(LEVEL_ID, ANOTHER_SCORE_VALUE, aUser);
+    }
+
+    @ThreadedAfter
+    public void after() {
+        Optional<Scores> scores = scoreRepo.findBy(aLevel());
+        assertThat(scores.isPresent()).isTrue();
+        assertThat(scores.get().contains(aScoreWith(aUser))).isTrue();
+        assertThat(scores.get().contains(aDifferentScoreWith(aUser))).isTrue();
+    }
+
+    @Test
+    public void testThreading() {
+        new AnnotatedTestRunner().runTests(this.getClass(), ScoreService.class);
+    }
+```
+
+This test will fail always, as is deterministic. As you can see testing concurrency is quite hard, and that's why I'm a supporter of modern frameworks that try to hide that hassle into platform or try to overcome the problem through immutable data. You can read more about it [here](http://felipefzdz.github.io/2015/01/02/concurrency-flavours/).
```

Comment:
This test will always fail, as it is deterministic. As you can see, testing concurrency is quite hard, and that's why I'm a supporter of modern frameworks that try to hide that hassle into a platform or overcome the problem through immutable data. You can read more about it [here](http://felipefzdz.github.io/2015/01/02/concurrency-flavours/).